test/extended/openstack/servers.go

```diff
@@ -157,8 +122,49 @@ var _ = g.Describe("[sig-installer][Suite:openshift/openstack] The OpenStack pla
 	})
 })
 
-// listWorkerMachineSets lists all worker machineSets
-func listWorkerMachineSets(ctx context.Context, dc dynamic.Interface) ([]objx.Map, error) {
+/*
+Checks that the Openstack created instances are as specified by the MachineSets.
+This function will be used for both checking workers and masters
+*/
+func commonCheckMachineSets(ctx context.Context, dc dynamic.Interface, computeClient *gophercloud.ServiceClient, machineSets []objx.Map) {
+	for _, machineSet := range machineSets {
+		nameMachineSet := machineSet.Get("metadata.name").String()
+		g.By(fmt.Sprintf("Getting the specs from the machineset %q", nameMachineSet))
+		replicasMachineSet, _ := strconv.Atoi(machineSet.Get("spec.replicas").String())
+		flavorMachineSet := machineSet.Get("spec.template.spec.providerSpec.value.flavor").String()
+		imageMachineSet := machineSet.Get("spec.template.spec.providerSpec.value.image").String()
+		sgMachineSets := getSecurityGroupNames(machineSet)
+		g.By(fmt.Sprintf("Getting the machines specs created by the machineset %q", nameMachineSet))
+		machines, err := getMachinesFromMachineSet(ctx, dc, nameMachineSet)
+		o.Expect(machines, err).To(o.HaveLen(replicasMachineSet),
+			"Number of replicas not matching for machineset %q", nameMachineSet)
+
+		for _, machine := range machines {
+			machine_name := machine.Get("metadata.name")
+			machine_phase := machine.Get("status.phase").String()
+			g.By(fmt.Sprintf("Get machine status for machine %q", machine_name))
+			o.Expect(machine_phase).To(o.Equal("Running"), "machine status is %q instead of Running", machine_phase)
+			g.By(fmt.Sprintf("Gather Openstack attributes for machine %q", machine_name))
+			instance, err := servers.Get(computeClient, machine.Get("metadata.annotations.openstack-resourceId").String()).Extract()
+			o.Expect(err).NotTo(o.HaveOccurred(), "Error gathering Openstack info for machine %v", machine_name)
+			g.By(fmt.Sprintf("Compare specs with openstack attributes for machine %q", instance.Name))
+			instanceFlavor, err := flavors.Get(computeClient, fmt.Sprintf("%v", instance.Flavor["id"])).Extract()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(instanceFlavor.Name).To(o.Equal(flavorMachineSet), "Flavor not matching for instance %q", instance.Name)
+			// Instance don't reference an image when using root volumes
+			if instance.Image["id"] != nil {
+				instanceImage, err := images.Get(computeClient, fmt.Sprintf("%v", instance.Image["id"])).Extract()
+				o.Expect(err).NotTo(o.HaveOccurred())
+				o.Expect(instanceImage.Name).To(o.Equal(imageMachineSet), "Image not matching for instance %q", instance.Name)
+			}
+			instanceSgs := parseInstanceSgs(instance.SecurityGroups)
+			o.Expect(instanceSgs).To(o.Equal(sgMachineSets), "SGs not matching for %q", instance.Name)
+		}
+	}
+}
+
+// listMachineSetsByRole lists all machineSets based on role label
+func listMachineSetsByRole(ctx context.Context, dc dynamic.Interface, role string) ([]objx.Map, error) {
```

Comment:
(posting again because Github hid my previous comment)

This function should only deal with MachineSets. I don't even think it should bother checking the role, as it's always going to be "worker" AFAIK.

A separate function might return the CPMS.